pkg/api/scraped_content.go

```diff
@@ -0,0 +1,127 @@
+package api
+
+import (
+	"fmt"
+
+	"github.com/stashapp/stash/pkg/models"
+)
+
+// marshalScrapedScenes converts ScrapedContent into ScrapedScene. If conversion fails, an
+// error is returned to the caller.
+func marshalScrapedScenes(content []models.ScrapedContent) ([]*models.ScrapedScene, error) {
+	var ret []*models.ScrapedScene
+	for _, c := range content {
+		if c == nil {
+			ret = append(ret, nil)
+			continue
+		}
+
+		if s, ok := c.(*models.ScrapedScene); ok {
+			ret = append(ret, s)
+		} else {
+			return nil, fmt.Errorf("%w: cannot turn ScrapedContent into ScrapedScene", models.ErrConversion)
+		}
+	}
+
+	return ret, nil
+}
+
+// marshalScrapedPerformers converts ScrapedContent into ScrapedPerformer. If conversion
+// fails, an error is returned to the caller.
+func marshalScrapedPerformers(content []models.ScrapedContent) ([]*models.ScrapedPerformer, error) {
+	var ret []*models.ScrapedPerformer
+	for _, c := range content {
+		if c == nil {
+			ret = append(ret, nil)
+			continue
+		}
+
+		if p, ok := c.(*models.ScrapedPerformer); ok {
+			ret = append(ret, p)
+		} else {
+			return nil, fmt.Errorf("%w: cannot turn ScrapedContent into ScrapedPerformer", models.ErrConversion)
+		}
+	}
+
+	return ret, nil
+}
+
+// marshalScrapedGalleries converts ScrapedContent into ScrapedGallery. If
+// conversion fails, an error is returned.
+func marshalScrapedGalleries(content []models.ScrapedContent) ([]*models.ScrapedGallery, error) {
+	var ret []*models.ScrapedGallery
+	for _, c := range content {
+		if c == nil {
+			ret = append(ret, nil)
+			continue
+		}
+
+		if g, ok := c.(*models.ScrapedGallery); ok {
+			ret = append(ret, g)
+		} else {
+			return nil, fmt.Errorf("%w: cannot turn ScrapedContent into ScrapedGallery", models.ErrConversion)
+		}
+	}
+
+	return ret, nil
+}
+
+// marshalScrapedMovies converts ScrapedContent into ScrapedMovie. If conversion
+// fails, an error is returned.
+func marshalScrapedMovies(content []models.ScrapedContent) ([]*models.ScrapedMovie, error) {
+	var ret []*models.ScrapedMovie
+	for _, c := range content {
+		if c == nil {
+			ret = append(ret, nil)
+			continue
+		}
+
+		if m, ok := c.(*models.ScrapedMovie); ok {
+			ret = append(ret, m)
+		} else {
+			return nil, fmt.Errorf("%w: cannot turn ScrapedConetnt into ScrapedMovie", models.ErrConversion)
+		}
+	}
+
+	return ret, nil
+}
+
+// marshalScrapedPerformer will marshal a single performer
+func marshalScrapedPerformer(content models.ScrapedContent) (*models.ScrapedPerformer, error) {
+	p, err := marshalScrapedPerformers([]models.ScrapedContent{content})
+	if err != nil {
+		return nil, err
+	}
+
+	return p[0], nil
```

Comment:
In theory we could have a `p` slice here with the first element being `nil` and then some more `actual` performers.
Do we want/need to check for that? The same goes for all the marshalScrape single functions

Reply:
In practice this function is to facilitate the existing GraphQL fields. Many of those don't do any kind of multiplicity and only ever returns a single value. If the first value is nil, we want to return nil.

For the multiple case, if I query for [X, Y, Z] and Y fails, I expect [{...Xres...}, null, {...Yres...}] in the GraphQL. I.e., that the order is preserved. This works for the multiplicity-variants.

But it is arguably stronger just to return `[ScrapedContent]` in the GraphQL and use GraphQL to do this conversion for us. It's just that it won't work for existing fields.